README.md
# JSONM
[![Coverage Status](https://coveralls.io/repos/github/repetere/jsonm/badge.svg?branch=main)](https://coveralls.io/github/repetere/jsonm?branch=main) [![Build, Test, Release](https://github.com/repetere/jsonm/actions/workflows/release.yml/badge.svg)](https://github.com/repetere/jsonm/actions/workflows/release.yml)

## Description

**JSONM** is a module that creates Tensorflow models with JSON.

### Jump right in

JSONM is designed so software engineers and machine learning engineers can quickly build, train, save, and load Tensorflow based AI & ML Models. The JSONM Library includes a JSONM UMD module with batteries included so you can use JSONM in the browser without transpilers or any additional setup/configuration. The JSONM UMD is ideal for JAMstack Applications.

### Usage

The idea behind JSONM is to enable rapid model development. JSONM attempts to automate the data pre-processing and feature engineering needed for most modeling tasks. 

Data Scientists who are more comfortable finely tuning hyperparameters and controlling pre-processing, scaling and normalization of datasets can also configure JSONM to meet specific model requirements.

### What's included
JSONM currently supports the following models
- General purpose multivariate based predictions
- Multivariate linear regression based predictions
- General purpose multivariate classification based descriptions
- General purpose multivariate timeseries based forecasts
- General purpose content based recommendations *(coming soon)*
- General purpose cohort analysis *(coming soon)*
## Installation

```sh
$ npm i @jsonstack/jsonm
```

<link id="viewx-style-style-0" rel="stylesheet" type="text/css" href="https://unpkg.com/highlight.js@9.18.1/styles/darkula.css">

---
### JSONM Manual
 - [Getting Started](https://repetere.github.io/jsonm/manual/getting-started/index.html)
 - Working With Data
   - [Data Fetching](https://repetere.github.io/jsonm/manual/data-fetching/index.html)
   - [Feature Engineering](https://repetere.github.io/jsonm/manual/feature-engineering/index.html)
 - Working With Models
   - [Model Training](https://repetere.github.io/jsonm/manual/model-training/index.html)
   - [Model Evaluation](https://repetere.github.io/jsonm/manual/model-evaluation/index.html)
   - [Model Predictions](https://repetere.github.io/jsonm/manual/model-predictions/index.html)
   - [Saving and Loading Models](https://repetere.github.io/jsonm/manual/saving-and-loading-models/index.html) 
 - [Advanced Topics](https://repetere.github.io/jsonm/manual/advanced-topics/index.html)
   - [JSONM & JML Spec](https://repetere.github.io/jsonm/manual/spec/index.html)
   - [Examples](https://repetere.github.io/jsonm/manual/examples/index.html)
   - [Full API Docs](https://repetere.github.io/jsonm/)
---

### Basic Usage
```typescript
import * as tf from '@tensorflow/tfjs-node';
import { getModel, setBackend, } from '@jsonstack/jsonm';

//set tensorflow
setBackend(tf);

//Iris Dataset e.g from https://raw.githubusercontent.com/repetere/modelx-model/master/src/test/mock/data/iris_data.csv
const type = 'ai-classification';
const dataset = [
  {
    "sepal_length_cm": 5.1,
    "sepal_width_cm": 3.5,
    "petal_length_cm": 1.4,
    "petal_width_cm": 0.2,
    "plant": "Iris-setosa",
  },
//  ...
  {
    "sepal_length_cm": 7.0,
    "sepal_width_cm": 3.2,
    "petal_length_cm": 4.7,
    "petal_width_cm": 1.4,
    "plant": "Iris-versicolor",
  },
  // ...
  {
    "sepal_length_cm": 5.9,
    "sepal_width_cm": 3.0,
    "petal_length_cm": 5.1,
    "petal_width_cm": 1.8,
    "plant": "virginica",
  }
]
const inputs = ['sepal_length_cm','sepal_width_cm','petal_length_cm','petal_width_cm', ];
const outputs = [ 'plant',];
const on_progress = ({ completion_percentage, loss, epoch, status, logs, defaultLog, }) => { 
  console.log({ completion_percentage, loss, epoch, status, logs, defaultLog, });
}
const IrisModel = await getModel({
  type,
  dataset,
  inputs,
  outputs,
  on_progress,
}); 
await IrisModel.trainModel()
const predictions = await IrisModel.predictModel({ 
  prediction_inputs:[
    { sepal_length_cm: 5.1, sepal_width_cm: 3.5, petal_length_cm: 1.4, petal_width_cm: 0.2, },
    { sepal_length_cm: 5.9, sepal_width_cm: 3.0, petal_length_cm: 5.1, petal_width_cm: 1.8, },
  ],
}); // => [ { plant:'Iris-setosa' }, { plant:'Iris-virginica' }, ]

```

## Example ##
<iframe width="100%" height="500" src="https://jsfiddle.net/yawetse/4ph1vwes/21/embedded/result,js,html,css,resources/dark/" allowfullscreen="allowfullscreen" allowpaymentrequest frameborder="0"></iframe>

### Development

Note *Make sure you have typescript installed*

```sh
$ npm i -g typescript 
```

For generating documentation

```sh
$ npm run doc
```

### Notes

Check out [https://repetere.github.io/jsonm/](https://repetere.github.io/jsonm/) for the full jsonm Documentation

### Testing

```sh
$ npm test
```

### Contributing

Fork, write tests and create a pull request!

License

----

MIT

CHANGELOG.md
## [2.2.1](https://github.com/repetere/jsonm/compare/v2.2.0...v2.2.1) (2022-08-04)


### Bug Fixes

* added getModel validation with no inputs/outputs ([411010b](https://github.com/repetere/jsonm/commit/411010b43b66035059dc42b315fe47e79587004d))

# [2.2.0](https://github.com/repetere/jsonm/compare/v2.1.0...v2.2.0) (2022-07-31)


### Features

* adding support to generate models from google sheets ([4da3694](https://github.com/repetere/jsonm/commit/4da369451688971cfdd9ba3f3b16a6e3f5f2135c))

# [2.1.0](https://github.com/repetere/jsonm/compare/v2.0.0...v2.1.0) (2022-07-31)


### Features

* fixed browser build with updated dependencies ([6250745](https://github.com/repetere/jsonm/commit/6250745728cc96f81507774b2f3558cadc3dacbb))

# [2.0.0](https://github.com/repetere/jsonm/compare/v1.1.0...v2.0.0) (2022-07-26)


### Bug Fixes

* moving jsonstack-data to dependencies ([771101a](https://github.com/repetere/jsonm/commit/771101a4aef10915e5bf9670bbeb7faf6ca0d647))
* pulling out scikit and tensorflow ([fe8a123](https://github.com/repetere/jsonm/commit/fe8a1239c7b26e92b56d846935e52ed5ae6ebb5f))
* reference to dependencies ([0411116](https://github.com/repetere/jsonm/commit/0411116a8029968fdcff998bbab6cbd6d057c93f))
* updated jsonstack-model ([290120e](https://github.com/repetere/jsonm/commit/290120e949ba86a6d1084d16d061b0638b662950))
* updated jsonstack-model ([0b13cfc](https://github.com/repetere/jsonm/commit/0b13cfc4f651ffa51350916b6603be9e754bf5b2))
* updating dependencies ([c1070a1](https://github.com/repetere/jsonm/commit/c1070a1e8e9cb464c6dddac056ab1da7f3c0f866))


### Features

* added auto splitting datasets into training and prediction data ([0afba45](https://github.com/repetere/jsonm/commit/0afba45d2bc4c09300e037a5f9cde5ce3bd3e598))
* migrated builds to esbuild ([664d58e](https://github.com/repetere/jsonm/commit/664d58e73340850027c42da66c3e8bd98d61e151))

# [1.1.0](https://github.com/repetere/jsonm/compare/v1.0.4...v1.1.0) (2021-04-22)


### Features

* exposing feature methors ([08dd98a](https://github.com/repetere/jsonm/commit/08dd98afcc08852ca55c5c0efd63b8f0dbe0ab7d))
* exposing feature methors ([13e1f81](https://github.com/repetere/jsonm/commit/13e1f81c8f94e49cffaec24ca391c976894807ee))

## [1.0.4](https://github.com/repetere/jsonm/compare/v1.0.3...v1.0.4) (2021-04-17)


### Bug Fixes

* adding for basic prediction inputs ([97f4542](https://github.com/repetere/jsonm/commit/97f4542ca9574a0d6494beef8b6b3c12de8299cf))

## [1.0.3](https://github.com/repetere/jsonm/compare/v1.0.2...v1.0.3) (2021-04-07)


### Bug Fixes

* basic forecasting working ([a2c0b05](https://github.com/repetere/jsonm/commit/a2c0b0590e8dd04072561110a23682245aa8044a))

## [1.0.2](https://github.com/repetere/jsonm/compare/v1.0.1...v1.0.2) (2021-04-03)


### Bug Fixes

* publishing to outsights ([01061fc](https://github.com/repetere/jsonm/commit/01061fc0ba9e5d0c22bc7e8c9744aecfbf8aa2f8))

## [1.0.1](https://github.com/repetere/jsonm/compare/v1.0.0...v1.0.1) (2021-04-03)


### Bug Fixes

* **jsonm:** fixed declarative model parameters ([c74de3a](https://github.com/repetere/jsonm/commit/c74de3a06b23098f3fbf8a96f7dc143f314d4add))

# 1.0.0 (2021-04-02)


### Bug Fixes

* **ci:** missing node modules ([cd148e5](https://github.com/repetere/jsonm/commit/cd148e57d442d3e96479b5b7282a423426ea71bb))
* updated dependencies ([056ec3c](https://github.com/repetere/jsonm/commit/056ec3c6d681df8a075ff053d63ded7948fcbcec))


src/util.ts
import * as JSONStackDataTypes from '@jsonstack/data/src/DataSet';
// ts-node -O {\"module\":\"commonjs\"}  src/util.ts
import * as features from './features';
import { getParsedDate, } from './constants';
import {faker} from '@faker-js/faker';
faker.seed(0);
export type Fake = {
  [index: string]: any;
}
export const Faker: Fake = { ...faker };

export function randomNumber(min:number, max:number) {  
  return Math.random() * (max - min) + min; 
}
export function generateNumberRange(start: number, end: number): number[]{
  return [start].reduce((result:number[], val:number) => { 
    for (let i=val; i < (end+1); i++){
      result.push(i);
    }
    return result;
  }, []);
}
export function getDatum(customDate?: Date, customTransation: { amount?: number; late_payments?: boolean; } = {}) {
  const transaction = {
    amount: Faker.finance.amount(),
    late_payments: Faker.datatype.boolean(),
  };
  const { amount, late_payments = true, } = customTransation;
  transaction.amount = parseFloat(amount || transaction.amount);
  transaction.late_payments = late_payments;
  const date = customDate||Faker.date.between(new Date('2020-01-15'), new Date('2020-04-10'));
  const parsedDate = getParsedDate(date);
  return {
    ...transaction,
    ...parsedDate,
    date,
  }
}
export function getData(items: number) {
  return generateNumberRange(0, items-1).map(()=>getDatum());
}
export const timeseriesSort = (a: JSONStackDataTypes.Datum, b: JSONStackDataTypes.Datum) => a.date.valueOf() - b.date.valueOf();

export function getMockTimeseries() {
  const timeseriesData = [
    getDatum(new Date('2020-04-04T00:00:00.000Z'),{amount:407, late_payments:true}),
    getDatum(new Date('2020-04-05T00:00:00.000Z'),{amount:408, late_payments:true}),
    getDatum(new Date('2020-04-06T00:00:00.000Z'),{amount:309, late_payments:false}),
    getDatum(new Date('2020-04-07T00:00:00.000Z'),{amount:410, late_payments:true}),
    getDatum(new Date('2020-04-08T00:00:00.000Z'),{amount:411, late_payments:true}),
    getDatum(new Date('2020-04-09T00:00:00.000Z'),{amount:412, late_payments:true}),
    getDatum(new Date('2020-04-10T00:00:00.000Z'),{amount:313, late_payments:false}),
    getDatum(new Date('2020-04-11T00:00:00.000Z'),{amount:414, late_payments:true}),
    getDatum(new Date('2020-04-12T00:00:00.000Z'),{amount:415, late_payments:true}),
    getDatum(new Date('2020-04-13T00:00:00.000Z'),{amount:416, late_payments:true}),
    getDatum(new Date('2020-04-14T00:00:00.000Z'),{amount:317, late_payments:false}),
    getDatum(new Date('2020-04-15T00:00:00.000Z'),{amount:318, late_payments:false}),
    getDatum(new Date('2020-04-16T00:00:00.000Z'),{amount:419, late_payments:true}),
    getDatum(new Date('2020-04-17T00:00:00.000Z'),{amount:420, late_payments:true}),
    getDatum(new Date('2020-04-18T00:00:00.000Z'),{amount:421, late_payments:true}),
    getDatum(new Date('2020-04-19T00:00:00.000Z'),{amount:322, late_payments:false}),
    getDatum(new Date('2020-04-20T00:00:00.000Z'),{amount:423, late_payments:true}),
    getDatum(new Date('2020-04-21T00:00:00.000Z'),{amount:324, late_payments:false}),
    getDatum(new Date('2020-04-22T00:00:00.000Z'),{amount:325, late_payments:false}),
    getDatum(new Date('2020-04-23T00:00:00.000Z'),{amount:426, late_payments:true}),
  ].sort(timeseriesSort);
  const independent_variables = [
    // 'type',
    'late_payments',
    'month',
    'day'
  ];
  const dependent_variables = ['amount'];
  const input_independent_features = [
    {
      feature_field_name: 'late_payments',
      feature_field_type: features.AutoFeatureTypes.TEXT,
    },
    // {
    //   feature_field_name: 'type',
    //   feature_field_type: features.AutoFeatureTypes.TEXT,
    // },
    {
      feature_field_name: 'month',
      feature_field_type: features.AutoFeatureTypes.TEXT,
    },
    {
      feature_field_name: 'day',
      feature_field_type: features.AutoFeatureTypes.TEXT,
    },
  ];
  const prediction_inputs = [
    getDatum(new Date('2020-04-24T00:00:00.000Z'),{ late_payments:false}),
    getDatum(new Date('2020-04-25T00:00:00.000Z'),{ late_payments:false}),
    getDatum(new Date('2020-04-26T00:00:00.000Z'),{ late_payments:true}),];
  return {
    timeseriesData,
    independent_variables,
    dependent_variables,
    input_independent_features,
    prediction_inputs,
  };
}

export function getMockRegression() {
  const data = [
    { input_1: 1, input_2: 10, input_3: 100, ignored_1: 1, output_1: 1000, output_2: 10000, },
    { input_1: 2, input_2: 20, input_3: 200, ignored_1: 2, output_1: 2000, output_2: 20000, },
    { input_1: 3, input_2: 30, input_3: 300, ignored_1: 3, output_1: 3000, output_2: 30000, },
    { input_1: 4, input_2: 40, input_3: 400, ignored_1: 4, output_1: 4000, output_2: 40000, },
    { input_1: 5, input_2: 50, input_3: 500, ignored_1: 5, output_1: 5000, output_2: 50000, },
    { input_1: 6, input_2: 60, input_3: 600, ignored_1: 6, output_1: 6000, output_2: 60000, },
    { input_1: 7, input_2: 70, input_3: 700, ignored_1: 7, output_1: 7000, output_2: 70000, },
    { input_1: 8, input_2: 80, input_3: 800, ignored_1: 8, output_1: 8000, output_2: 80000, },
    { input_1: 9, input_2: 90, input_3: 900, ignored_1: 9, output_1: 9000, output_2: 90000, },
    { input_1: 11, input_2: 110, input_3: 1100, ignored_1: 11, output_1: 11000, output_2: 110000, },
    { input_1: 12, input_2: 120, input_3: 1200, ignored_1: 12, output_1: 12000, output_2: 120000, },
    { input_1: 13, input_2: 130, input_3: 1300, ignored_1: 13, output_1: 13000, output_2: 130000, },
    { input_1: 14, input_2: 140, input_3: 1400, ignored_1: 14, output_1: 14000, output_2: 140000, },
    { input_1: 15, input_2: 150, input_3: 1500, ignored_1: 15, output_1: 15000, output_2: 150000, },
    { input_1: 16, input_2: 160, input_3: 1600, ignored_1: 16, output_1: 16000, output_2: 160000, },
    { input_1: 17, input_2: 170, input_3: 1700, ignored_1: 17, output_1: 17000, output_2: 170000, },
    { input_1: 18, input_2: 180, input_3: 1800, ignored_1: 18, output_1: 18000, output_2: 180000, },
    { input_1: 19, input_2: 190, input_3: 1900, ignored_1: 19, output_1: 19000, output_2: 190000, },
  ];
  const independent_variables = [
    'input_1',
    'input_2',
    'input_3',
  ];
  const dependent_variables = [
    'output_1',
    'output_2',
  ];
  const prediction_inputs = [
    { input_1: 5, input_2: 50, input_3: 500, ignored_1: 5, },
    { input_1: 10, input_2: 100, input_3: 1000, ignored_1: 10, },
    { input_1: 15, input_2: 150, input_3: 1500, ignored_1: 15, },
    { input_1: 20, input_2: 200, input_3: 2000, ignored_1: 20, },
    { input_1: 40, input_2: 400, input_3: 4000, ignored_1: 40, },
  ];
  return {
    data,
    independent_variables,
    dependent_variables,
    prediction_inputs,
  };
}

export function getMockClassification() {
  const data = [
    { walking_noise_level: 2, primary_sound: 'yip', secondary_sound: 'growl', weight: 10, ear_style: 'pointy', animal: 'dog', },
    { walking_noise_level: 4, primary_sound: 'bark', secondary_sound: 'growl', weight: 25, ear_style: 'floppy', animal: 'dog', },
    { walking_noise_level: 1, primary_sound: 'meow', secondary_sound: 'pur', weight: 12, ear_style: 'pointy', animal: 'cat', },
    { walking_noise_level: 8, primary_sound: 'bark', secondary_sound: 'growl', weight: 50, ear_style: 'pointy', animal: 'dog', },
    { walking_noise_level: 2, primary_sound: 'meow', secondary_sound: 'growl', weight: 15, ear_style: 'pointy', animal: 'cat', },
    { walking_noise_level: 2, primary_sound: 'yip', secondary_sound: 'pur', weight: 15, ear_style: 'pointy', animal: 'cat', },
    { walking_noise_level: 2, primary_sound: 'meow', secondary_sound: 'pur', weight: 15, ear_style: 'pointy', animal: 'cat', },
    { walking_noise_level: 1, primary_sound: 'yip', secondary_sound: 'growl', weight: 25, ear_style: 'floppy', animal: 'dog', },
    { walking_noise_level: 2, primary_sound: 'pur', secondary_sound: 'growl', weight: 15, ear_style: 'pointy', animal: 'cat', },
  ];
  const prediction_inputs = [
    { walking_noise_level: 4, primary_sound: 'bark', secondary_sound: 'growl', weight: 25, ear_style: 'floppy', },
    { walking_noise_level: 1, primary_sound: 'meow', secondary_sound: 'pur', weight: 9, ear_style: 'pointy', },
  ];
  const independent_variables = [
    'walking_noise_level',
    'primary_sound',
    'secondary_sound',
    'weight',
    'ear_style',
  ];
  const dependent_variables = [
    'animal',
  ];
  return {
    data,
    independent_variables,
    dependent_variables,
    prediction_inputs,
  }
}

export const dataset0 = [{ num:1, some: 'val dataset0 1' }, { num:1, some: 'val dataset0 2' }, { num:1, some: 'val dataset0 3' }];
export const dataset1 = [{ num:1, some1: 'val dataset1' }, { num:1, some1: 'val dataset1' }, { num:1, some1: 'val dataset1' }];
export const dataset2 = {
  name: 'custom_name_2',
  data: [{ num:1, some2: 'val custom_name_2' }, { num:1, some2: 'val custom_name_2' }, { num:1, some2: 'val custom_name_2' }],
};
export const dataset3 = {
  data: [{ num:1, some3: 'val dataset3 1' }, { num:1, some3: 'val dataset3' }, { num:1, some3: 'val dataset3' }],
};
export const dataset4 = {
  name: 'async dataset',
  _data_promise: () => Promise.resolve([{ num:1, some4: 'val dataset4' }, { num:1, some4: 'val dataset4' }, { num:1, some4: 'val dataset4' }]),
};

  /*
  describe('Single Value Timeseries Predictions', () => {
    it('should forecast the number of passengers', async () => {
      const csvPath = path.join(__dirname, '../manual/media/example/tensorflowcsv/airline-trips-sales.csv');

      const airline_prediction_inputs = [
        {
          Month: '1960-01',
          Flights: 47,
          Stops: 4,
          Tickets: 417,
        },
        {
          Month: '1960-02',
          Flights: 31,
          Stops: 3,
          Tickets: 391,
        },
        {
          Month: '1960-03',
          Flights: 49,
          Stops: 4,
          Tickets: 419,
        },
        {
          Month: '1960-04',
          Flights: 41,
          Stops: 4,
          Tickets: 461,
        },
        {
          Month: '1960-05',
          Flights: 42,
          Stops: 4,
          Tickets: 472,
        },
        {
          Month: '1960-06',
          Flights: 55,
          Stops: 5,
          Tickets: 535,
        },
        {
          Month: '1960-07',
          Flights: 62,
          Stops: 6,
          Tickets: 622,
        },///
        {
          Month: '1960-08',
          Flights: 66,
          Stops: 6,
          Tickets: 606,
        },
        {
          Month: '1960-09',
          Flights: 58,
          Stops: 5,
          Tickets: 508,
        },
        {
          Month: '1960-10',
          Flights: 41,
          Stops: 4,
          Tickets: 461,
        },
        {
          Month: '1960-11',
          Flights: 30,
          Stops: 3,
          Tickets: 390,
        },
        {
          Month: '1960-12',
          Flights: 42,
          Stops: 4,
          Tickets: 432,
        },
        {
          Month: '1961-01',
          Flights: 47,
          Stops: 4,
          Tickets: 427,
        },
        {
          Month: '1961-02',
          Flights: 41,
          Stops: 4,
          Tickets: 401,
        },
        {
          Month: '1961-03',
          Flights: 49,
          Stops: 4,
          Tickets: 429,
        },
      ];

      const independentVariables = [
        'Passengers',
      ];
      const dependentVariables = [
        'Passengers',
      ];
      const airlineColumns = [].concat(independentVariables, dependentVariables);
      const airlinetrainning_feature_column_options = airlineColumns
        .reduce((result, val) => {
          result[val] = ['scale', 'standard',];
          return result;
        }, {});
      const airlineData = await csv.loadCSV(csvPath);
      // console.log(airlineData.slice(100));
      const timeseriesTestEnvParameters = {
        modelDocument: {
          model_configuration: {
            model_type: 'ai-fast-forecast',
            // model_type:'ai-classification',
            model_category: 'timeseries',
          },
        },
        
      };
      const timeseriesModelTest = new ModelX({
        model_type:ModelTypes.FAST_FORECAST,
        training_options: {
          fit: {
            epochs: 50,
            batchSize: 1,
          },
          stateful: true,
          // lookBack: 3,
        },
        trainingData: airlineData,
        training_feature_column_options: airlinetrainning_feature_column_options,
        x_independent_features: independentVariables,
        y_dependent_labels: dependentVariables,
        prediction_timeseries_date_feature: 'Month',
        prediction_timeseries_start_date: '1960-01',
        prediction_timeseries_end_date: '1962-12',
        // y_raw_dependent_labels:rawDependentVariables,
      });
  
      const predictions = await timeseriesModelTest.predictModel({
        retrain: true,
        // cross_validate_training_data: false,
        // fixedModel: false,
        prediction_inputs: airline_prediction_inputs,
      });
      // console.log({ predictions });
      expect(predictions).toBeInstanceOf('array');
      expect(predictions.length).toBe(36);
    });
  });
  */